src/parser.rs

```diff
@@ -0,0 +1,242 @@
+use crate::expr::{BinaryOp, Expr, Literal, UnaryOp};
+use crate::parse_error::ParseError;
+use crate::token::token;
+use crate::token::token::Token;
+use crate::token::token_type::TokenType;
+
+#[derive(Default)]
+pub struct Parser<'a> {
+    pub tokens: Vec<Token<'a>>,
+    pub current: usize,
+}
+
+impl<'a> Parser<'a> {
+    pub fn parse(self) -> u8 {
+        todo!();
+    }
+
+    fn expression(&mut self) -> Result<Expr, ParseError> {
+        return self.equality();
+    }
+
+    fn equality(&mut self) -> Result<Expr, ParseError> {
+        let mut expr = self.comparison()?;
+        while self.match_types(vec![TokenType::BANG_EQUAL, TokenType::EQUAL_EQUAL]) {
+            let operator = self.previous();
+            let maybe_binary_operator = Parser::token_to_binary_operator(operator);
+
+            let right = self.comparison()?;
```

Comment:
I was getting really weird borrow errors here. I was looking at my IDE and it wasn't really explaining the issue well. Looking at the terminal output pointed me in the right direction. 

I previously had it like this:

```suggestion
            let operator = self.previous();
            let right = self.comparison()?;
            let maybe_binary_operator = Parser::token_to_binary_operator(operator);

```

And the order really mattered? I should make sure to use the `operator` variable before calling the other method since the other method couldn't take ownership of `self`(?) while `operator` was still being used here in `equality`?

Reply:
If you're borrowing some or all of a struct (e.g. by calling `let right = self.comparison()`) then you can't "give" the struct to a mutable function (I see `previous` is `fn previous(&mut self) -> &Token`). This is because the function which has the mutable reference could change the data from under you. So you've got some options: 

 - Change the call order (like you did) so that mutations happen before reads.
 - Change functions to avoid references, and instead copy or clone data (less memory efficient but frequently worth doing)
 - Try to find ways to avoid `&mut`. It looks like `previous()` doesn't actually _need_ `&mut self` so maybe try changing it to `&self` and see if you can change the order without compile failures.